### ISSUE TITLE
Implement support for PUSH notifications for Android (XEP-0357)

### ISSUE THREAD

**Issue by matwey** [redsolution/xabber-android#529]
Hello,

EJabberd has support of XEP-0357 to allow clients to handle Android PUSH notifications in order to save power and suspend network connection.
For instance, gcm.jabber.ru is available
Would be great to have support of the feature in Xabber.

**Comment by Flowdalic** [comment]
> in order to save power and suspend network connection.

Using GCM and XEP-0357 on Android does not have any advantage  over an properly designed and implemented XMPP client for Android.

**Comment by zinid** [comment]
> Using GCM and XEP-0357 on Android does not have any advantage over an properly designed and implemented XMPP client for Android

Any proofs?

**Comment by Flowdalic** [comment]
> Any proofs?
It does not matter if you transfer the data via GCM or via XMPP, neither can transfer the same amount of data with the same realtime semantics in a more "battery friendly" way then the other. Both use the same radio device and usually the same network protocols. An unnecessary active radio is what's causing unnecessary battery consumption. What matters is taking the right decision *when* and *if* you transfer data from the server to the client and vice versa. And GCM can't help you with that. XEP-357 sole reason for existence is to support platforms which (usually) don't allow long standing TCP connection, e.g. iOS.

**Comment by zinid** [comment]
@Flowdalic I'm not convinced. I would rather see benchmarks of an XMPP application with pushed enabled and push disabled. Otherwise those are only speculations.

**Comment by bangarharshit** [comment]
@Flowdalic In marshmallow app standby will not allow long standing TCP connections - http://developer.android.com/about/versions/marshmallow/android-6.0-changes.html#behavior-power.

However there are few concerns for using GCM:
1. I found GCM highly unreliable. http://eladnava.com/google-cloud-messaging-extremely-unreliable/ . Though the article claims they reduced the ping time to 8 minutes but on my phone it is still 28 minutes.
2. The cost for establishing connection is high. Without something like xep-0305 quick start the real time aspect is lost especially for something like jingle sessions. You still need to connect inside wakelock or alarmmanager (async). For jingle you need to make sure that wakelock is up until the session connection is established.

For power there is just one ping central to android device if GCM is used. If alarm manager with setInexact is used it shouldn't affect much since radio will be up in parallel. I can't think of much changes otherwise.

**Comment by andrewnenakhov** [comment]
@Flowdalic FCM DO offer advantages over XMPP. Android OS makes sure that app can be in suspended state and be woken by push notification. This way XMPP client can cease to receive lot's of presence stanzas, and ultimately save some data. However, proper implementation has some tricky questions.

**Comment by Flowdalic** [comment]
> Android OS makes sure that app can be in suspended state and be woken by push notification. 

Your application will also be woken up by an incoming XMPP stanza if it is whitelisted from doze mode and if you are not running Android O preview.

**Comment by andrewnenakhov** [comment]
@Flowdalic app won't be woken by incoming XMPP stanza if it does not maintain XMPP connection, and it can't maintain it if it's suspended. Oops.

**Comment by Flowdalic** [comment]
> @Flowdalic app won't be woken by incoming XMPP stanza if it does not maintain XMPP connection, and it can't maintain it if it's suspended. Oops.

Define "suspended". A background (sticky) service part of an app which is whitelisted from doze mode is perfectly capable maintaining an TCP connection, even while the device is in (deep) sleep.

**Comment by vitalyster** [comment]
@Flowdalic 
> if it is whitelisted
> if you are not running Android O

Google forbids background tcp services now. Period. Stop this FUD about "properly designed xmpp client/service". Google already designed background tcp service (using xmpp in some cases, but it doesn't matter). Client apps must rely on this service and should not reinvent the wheel. Please, design "proper services" in your own Google competing platform, please stop disturb client developers writing good chat client for Android with your shitty XSF political statements

**Comment by andrewnenakhov** [comment]
@vitalyster ahem. Actually Xabber does works somehow despite 'google forbids background tcp services', though Google does cause some problems with maintaining background connection in Android. Nothing too unavoidable. Their reasoning is understandable, however, their offered solution (FCM) for messaging is not ideal and does not cover all usecases - like a functioning XMPP client.

**Comment by vitalyster** [comment]
@andrewnenakhov 
>  Google does cause some problems
No, Google just want you to not use tcp in background and it is not necessary for "functioning XMPP client". Client should be suspended as any other application.

**Comment by andrewnenakhov** [comment]
@vitalyster maybe you should just write an XMPP client that works like you describe and show us how it should be done?

**Comment by vitalyster** [comment]
@andrewnenakhov it is described in XEP-0357.

**Comment by andrewnenakhov** [comment]
@vitalyster so if you've read XEP-0357 you must have read that for push notifications to work it needs a **user agent** -  a service running locally on the user's device which receives push notifications and delivers them to the appropriate application.

Then, this user agent -"mapped as App Client"  "becomes the XMPP User Agent", then, "The App Client performs any necessary bootstrapping and registration for its preferred push service. - and, you guess what, it has to RUN LOCALLY TO RECEIVE NOTIFICATIONS, and it can't receive any push notification if it's suspended. Oops. 

If you do not agree, then come on, do an implementation of XEP-0357 and show us how it should be done.

**Comment by vitalyster** [comment]
@andrewnenakhov take a look at `GcmListenerService`/`FirebaseMessagingService` - that is what "user agent" part will be, and it will wake up client on FCM events

**Comment by andrewnenakhov** [comment]
@vitalyster *facepalm*. Re-read conversation from the beginning please. 

> No, Google just want you to not use tcp in background and it is not necessary for "functioning XMPP client". Client should be suspended as any other application.

and 

> take a look at GcmListenerService/FirebaseMessagingService - that is what "user agent" part will be, and it will wake up client on FCM events

Actually everyone knows about FCM and how it works, it's pros and cons. However, your statement 

> Client apps must rely on this service and should not reinvent the wheel.

is just wrong. Not everyone wants to lock in Google's proprietary ecosystem, not every Android out has Google Services, third-party servers can't send FCM to wake Xabber without a proxy run by Xabber developers, and many more tricky issues.

**Comment by vitalyster** [comment]
> Not everyone wants to lock in Google's proprietary ecosystem

Yes. Amazon/Nokia/[name other brand] provides their own ecosystem on top of Android, and their own "notifications" implementation. And applications designed e.g. for Amazon Store are support Amazon Device Messaging too.
Designing own notifications service for "geeks" which just hate any "proprietary ecosystem" is completely different task and NOT related to *messenger* development.
Anyway, you can: check for google services presence and register device in FCM, (check for Amazon DM/etc), otherwise start your own homemade tcp background service. 
Running tcp service on Google ecosystem you are just make ruined user experience and 99.9% normal users will not use this "tricky" app, which requires "whitelisting" and other tinkering.

**Comment by andrewnenakhov** [comment]
@vitalyster actually many users want to use their Android devices just like a regular computer and run whatever software they feel like using in a way they want, not how Google overlords allow them to do. XMPP by it's nature and design requires a client to maintain active TCP connection to server. Xabber does just that, with more or less success on most devices. 

Using FCM will allow some users to gain better battery life from their devices by letting Xabber be suspended until waking FCM push comes, however, FCM (and APNS too) are known to be much less reliable then active connection. When C2DM came around it sometimes took hours to reach device. GCM/FCM these days usually do better, but still it is less reliably 'instant' than messages in active XMPP stream. Since Xabber with 2 very active accounts uses now 1-2% of battery life from my device, I'm rather happy to use more battery power for faster messaging.

**Comment by vitalyster** [comment]
> many users want to use their Android devices just like a regular computer

How many? 1% of the Android userbase? And you want to ignore 99% others? I've already provide a case how to satisfy "normies" and "geeks" at the same time - detect Google services presence on the device.

> XMPP by it's nature and design requires a client to maintain active TCP connection

That is not true and that is what I've named "XSF political statement". 
XMPP connection != user session. Yes, most of XMPP specifications/extensions assume it is the same, but is not acceptable in the mobile world, it is possible to break this and there are many specifications which defines user session abstract of underlying connection already (XEP-0198 session resuming at first).

> known to be much less reliable then active connection

This statement from XMPP nerds make me fun every time. So Whatsapp and friends used by billions of users works 'unreliable', while "classic" XMPP clients used by few "geeks" work "better", are you sure?

> uses now 1-2% of battery life from my device

... while sitting in front of computer on the wifi connection?

> faster messaging.

When client app is in foreground (user actively chatting) it's obvious to create TCP connection and XMPP connection fits perfectly for this active state of the app, but not for the background one.

**Comment by andrewnenakhov** [comment]
> ... while sitting in front of computer on the wifi connection?

No, with lot's of driving, chatting on the go, etc. I move alot and basically run my company with Xabber. 

> This statement from XMPP nerds make me fun every time. So Whatsapp and friends used by billions of users works 'unreliable', while "classic" XMPP clients used by few "geeks" work "better", are you sure?

Yes, @vitalyster , unreliable. I've personally tested Whatsapp many years ago and sometimes it took HOURS (max time was like 6 hours) for message to reach phone and ring a notification on it. All because it relied on c2dm. It surely has it's merits, but they are totally not related to push notifications.

Anyway, you are free to implement any unnerdy XMPP client or service of your own, but I suggest you leave this issue tracker forever and let us XMPP nerds do what we think is right.

**Comment by vitalyster** [comment]
> many years ago

Things changing too fast, even Whatsapp is working faster noew, there are many alternatives, which works excellent too. While still no opensource replacement for these "proprietary services".

> let us XMPP nerds do what we think is right

> this issue tracker

This "proprietary" tracker? How you can use it? Real geeks not using proprietary services!

No problem, I understand you point. I just think you are the company who wants to dominate the world with opensource messaging solution, but you are playing "just for fun" with "dreaming android is a computer". Good luck! Will not bother you again

**Comment by licaon-kter** [comment]
2.6(557) server info says "push not supported"

How is this tested exactly?

**Comment by panpansh** [comment]
Please dont use google services, amazon etc. This guy dont know what is the word privacy.

**Comment by nponeccop** [comment]
An OS vendor notifications mechanism is supposed to be both power-saving and ensuring faster delivery compared to long-standing TCP connections. 

If this is confirmed by a benchmark  (of course it's a big *if*) then an optional mechanism should be implemented, which trades off some privacy for timely delivery and battery life. And users caring about privacy can then choose as they wish.

Note that there is more than one issue which a proper implementation must account for:

- NAT timeout
- power saving
- IP mobility

They all are _reportedly_ fixed by a third-party push notification service. So _multiple_ benchmarks are needed.

A good _starting_ benchmark would be sending messages at random large intervals (hours) to an idle phone ("lying on a shelf for hours") and noticing the delivery delays.

We can compare various messengers then to determine state of the art. E.g. WhatsApp and Wickr are both known NOT to have this delayed reception problem (I think it's irrelevant whether they use the vendor-provided service or implement their own "UDP smart polling" or whatever). So if Xabber performs as good as WhatsApp in practice, then it's a good reason to close this issue.

If Xabber performs statistically worse, then it's not an immediate reason to resort to GCM, Firebase or whatever is it's called now. We should attempt to fix this issue "normally". Only when fast delivery turns out not to be achievable by simple means, should this privacy-affecting opt-in feature be tried.
